src/main/scala/cognite/spark/v1/MetricsSource.scala

```diff
@@ -43,9 +62,53 @@ class MetricsSource {
           metrics
         }
       }
-    )
+
+    SparkEnv.get.metricsSystem.registerSource(source)
+
+    source
+  }
+
+  def deregisterMetricSource(key: String, source: Source): Unit = {
+    metricsMap.remove(key)
+    SparkEnv.get.metricsSystem.removeSource(source)
   }
 }
 
 // Singleton to make sure each metric is only registered once.
 object MetricsSource extends MetricsSource
+
+class ExpiringCounter(key: String, source: Source) extends Counter {
+  private var (deathTimer, deathTimerTask): (Timer, TimerTask) = createTimer
+
+  private def createTimer = {
+    val task = new TimerTask {
+      override def run =
+        MetricsSource.deregisterMetricSource(key, source)
+    }
+
+    val timer = new Timer(key)
+    timer.schedule(task, (1 hour).toMillis)
+
+    (timer, task)
+  }
+
+  private def resetTimer() = {
+    deathTimerTask.cancel()
+    deathTimer.cancel()
+
+    val (timer, task) = createTimer
+    deathTimer = timer
+    deathTimerTask = task
+  }
+
+  override def inc(n: Long): Unit = {
+    resetTimer()
```

Comment:
some potential for perf degradation here, before it was more or less simple increment, now it is noticeable work per increment (a little less scary since it'd probably be batch-increment but still). Especially the recreation of Timer and its thread